docs/dev.md

```diff
@@ -0,0 +1,143 @@
+than 72 characters) then be as close as possible.
+
+For example:
+
+```
+meson: bumpt gtest requirement from 1.12 to 1.15
+
+gtest 1.13 fixes a bug on Windows that is affecting our CI, see
+google/googletest#1234.
+
+gtest 1.14 adds a new feature that allows for faster test execution that will be
+used in the next patch
+```
+
+Finally, if a commit fixes any github issues, they should add that,
+
+```
+Fixes: #7
+```
+
+### Merging and rebasing
+
+Rebasing should generally be preferred to merges. Rebases create a linear
+history that is easy to bisect over, and easy for a human to understand. In a
+project like cps-config that lacks large subsystems this is easy to maintain.
+
+Merges may be appropriate for large additions or refactors that have moved
+quickly and may not have a clean history. This should generally be unnecessary.
+Merges also do not go through CI, and as such it's possible to create a state
+where main before the merge builds successfully, and the topic branch before the
+merge builds successfully, but the merge commit itself is broken.
+
+Back merges (merging the main branch back into a topic branch) should never
+happen. This makes bisects particularly hard and the history very messy. A
+rebase should be done in this case
+
+## Editors
+
+Use [editorconfig](https://editorconfig.org). Many editors support editorconfig
+natively, others have plugins. Otherwise, configure your editor to use the same
+settings.
+
+## All files
+
+All files should have an [SPDX license
+header](https://spdx.dev/learn/handling-license-info/), along with the relevant
+copyright holder information, as long as they have a comment syntax. JSON files
+do not, and as such should not have such a header.
+
+For example, in a C or C++ file:
+```cpp
+// SPDX-License-Identifier: MIT
+// Copyright 2024 Copyright Holder
+```
+
+## Meson
+
+Use the following naming conventions:
+
+- external_dependencies (found by `dependency()` or `compiler.find_library()`)
+  should be prefixed with `dep_`, ex: `dep_fmt`
+- internal dependencies (`declare_dependency()`) should be prefixed with `idep_`
+- external programs (`find_program()`) should be prefixed with `prog_`
+
+use spaces around `:`, ex: `dependency('gtest', version : '>= 1.5')`
+
+always drop arguments to a new line when they do not fit on one:
+
+```meson
+project(
+  'cps-config',
+  'cpp',
+  version : '0.0.1',
+)
+```
+
+## C and C++
+
+Prefer C++ (`//`) comments to C-style (`/* */`) in general.
+
+C style comments may be used in the following situations:
+  - spec citations
+  - algorithm descriptions
+
+For example:
+
+```c++
+/* Do X before Y, unless Z.
+ *
+ * Per the CPS Spec (0.9.0):
+ *
+ *   When doing thing, X must be resolved before Y, exception when Z is true,
+ *   then Y must be done and X must not.
+ *
+ */
+void do_thing(bool x, bool y, bool z);
+```
+
+`#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
+`CPS_CONFIG_USE_UNDEFINED`
```

Comment:
```suggestion
`#defines` should be all uppercase, and be prefaced with `CPS_CONFIG_`. such as
`CPS_CONFIG_USE_UNDEFINED`
```

Well, this is subtle. If we're going to ship a `cps` library, the prefix would be `CPS_`. The executable could also use `CPS_*` because it's shipping in the same repo so collisions are unlikely-to-impossible. Or it could use `CPS_CONFIG_` if we want to treat `cps` and `cps-config` as two projects sharing one build system and repo.

I'm flexible, but I did want to point out that `CPS_CONFIG_FOOBAR` would be odd in a `cps/foobar.hpp`.